drivers: apic_timer: move to periodic mode

npitre opened the pull request:
This driver is impossible to make time-accurate using single-shot
mode. Time accuracy may be obtained only by using periodic mode, meaning
it is not tickless capable either. Let's simplify the code by only
supporting periodic mode and strip out the TSC stuff. Any hardware with
TSC capability should use the apic_tsc driver instead.

There is apparently some hardware out there with a TSC but without the
ability to set a deadline comparator. In such case it would be much
cleaner to add the ability to use the APIC ICR in one-shot mode as a 
fallback IRQ source to the apic_tsc driver instead of relying on the 
apic_timer driver.

npitre committed:
drivers/timer/apic_tsc: move to common code pattern

Let's replicate a common code pattern for this to be abstracted more
easily in the future. In addition to duplicating the correctness fixes
implemented in the ARM and RISC-V drivers, this eliminates a couple large
runtime divisions.

Signed-off-by: Nicolas Pitre <npitre@baylibre.com>

npitre committed:
drivers/timer/apic_timer: move to periodic mode

This driver is impossible to make time-accurate using single-shot
mode. Time accuracy may be obtained only by using periodic mode, meaning
it is not tickless capable either. Let's simplify the code by only
supporting periodic mode and strip out the TSC stuff. Any hardware with
TSC capability should now use the apic-tsc driver instead.

Signed-off-by: Nicolas Pitre <npitre@baylibre.com>

andyross commented on drivers/timer/apic_timer.c:
```diff
@@ -31,193 +27,63 @@ BUILD_ASSERT(!IS_ENABLED(CONFIG_SMP), "APIC timer doesn't support SMP");
-	k_spin_unlock(&lock, key);
-	sys_clock_announce(ticks);
-}
+BUILD_ASSERT(CYCLES_PER_TICK >= 1, "APIC timer: bad CYCLES_PER_TICK");
```
Might want a BUILD_ASSERT(!defined(CONFIG_TICKLESS_KERNEL)) given that support is being removed, for the benefit of any out-of-tree apps that might be using it.

andyross reviewed (commented):
Some notes, haven't had time for a full review.  No doubt lots of bikeshedding will happen over the feature removal, but just for the record this looks fine to me.

npitre committed:
drivers/timer/apic_tsc: use ICR as a fallback timeout event source

This adds support for the local APIC in one-shot mode as the timeout
event source for those cases where the CPU supports invariant TSC but
no TSC deadline capability. It is presented as another timer choice.
Existing Kconfig symbols were preserved to minimize board config
disturbance.

This hybrid approach was implemented kind of backward in the apic_timer
driver but it is far cleaner to carry this here.

Signed-off-by: Nicolas Pitre <npitre@baylibre.com>

npitre commented:
> No doubt lots of bikeshedding will happen over the feature removal

Fear no more. The bikeshed has retained all its features with the last push.  ;-)

andyross reviewed (dismissed):
Looks as clean as this hardware is as likely to get.

kwd-doodling commented on drivers/timer/apic_tsc.c:
```diff
@@ -39,63 +101,103 @@ static ALWAYS_INLINE uint64_t rdtsc(void)
+	sys_clock_announce(delta_ticks);
 }
 
 void sys_clock_set_timeout(int32_t ticks, bool idle)
```
one issue captured on ISH here, there's a chance that sys_clock_set_timeout() is called with ticks=0. Then ICR is programmed with zero, so the timer stops.

npitre replied:
Fixed.

kwd-doodling commented:
@npitre this patch does enhance apic_tsc.c, look good 👋
